### PR TITLE
New version of reflectometry `Auto` algorithms

### LINKED ISSUE #17429: Investigate if `ReflectometryReductionOne` can be refactored
We now have detailed workflow diagrams for `ReflectometryReductionOne` and `CreateTransmissionWorkspace`, and it looks like some of the steps are duplicate (for instance some `CropWorkspace` and `Rebin` steps) and could be avoided if the algorithm was refactored.

To make sure that we are refactoring it properly we should have better unit tests.

This is an umbrella issue (sub-issues can be created for each task if necessary) to:
- [x] Create a new version of `ReflectometryReductionOne` (as we may remove/introduce new properties) from the current version
- [x] Refactor the new version of `ReflectometryReductionOne`
- [x] Write unit tests that cover all the input properties
- [x] Discuss `CreateTransmissionWorkspace` with scientists to see if they are happy with the current version.

### LINKED ISSUE #18029: Update `ReflectometryReductionOneAuto` and `CreateTransmissionWorkspaceAuto`
Update these two algorithms, and possibly `SpecularReflectionPositionCorrect` according to changes in #17429.

### PR DESCRIPTION
**Pull request by raquelalvarezbanos**
New versions of some Reflectometry algorithms:

- `ReflectometryReductionOne` and `ReflectometryReductionOneAuto`.
- `CreateTransmissionWorkspace` and `CreateTransmissionWorkspaceAuto`.
- `SpecularReflectionPositionCorrect`

**To test:**

- Check that new algorithms have their own unit tests (and check that the new test files are testing the new version).
- Check that no functionality was added/removed in the old versions (i.e. check files `ReflectometryReductionOne`, `ReflectometryReductionOneAuto`, `CreateTransmissionWorkspace`, `CreateTransmissionWorkspaceAuto`, `SpecularReflectionPositionCorrect` and ensure code changes are minimal, both headers and source files). Only some tests were added.
- Check that each of the new algorithms comes with an .rst file. Build the documentation and make sure it is accurate.
- System test `RROAutoFunctionalityTests` was removed. The reason is that it uses a decorator, `construct_standard_algorithm()` that made impossible selecting the old version of the algorithm. Check that those tests have been moved to `ReflectometryReductionOneAutoTest`.
- Check that release notes were updated.
- Check that existing system tests are testing version 1 of the algorithms.
- Run this script: [ReflectometryReductionOneAuto_comparison.zip](https://github.com/mantidproject/mantid/files/708726/ReflectometryReductionOneAuto_comparison.zip). It is the script I've been using to compare versions 1 and 2 of `ReflectometryReductionOneAuto` (it also tests indirectly the other algorithms mentioned above). Plot the `IvsLam` workspaces generated by each version. They should be compatible within error bars and should be very similar (some differences are expected in some cases, such as POLREF data, specially if data are noisy).
- Code review: have a look at the new classes listed above (I also added a new version of `ReflectometryWorkflowBase` although it does not have unit tests).
- Check that the old interface is using version 1 of the algorithms (this will be changed in PR https://github.com/mantidproject/mantid/pull/18239)
- Check that the new interface is using version 2 of the algorithms. Test that the interface works reducing a few datasets. I've checked with the .tbl we have for INTER. Check that results look exactly the same as in master. Also, check that the parameters (such as the angle, Q min, Qmax, etc) are updated in the table when the reduction is complete.

Fixes #17429.
Fixes #18029.

Release notes: https://github.com/mantidproject/mantid/pull/18058/commits/b0bea6d9828431f5a53bd472268f3a9154a49a4a and https://github.com/mantidproject/mantid/pull/18058/commits/83ca77c53c6ea58ec76859237431ee9a95d089c7

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by raquelalvarezbanos** [comment]
Max is happy with the new versions. I am now waiting for feedback from Jos.

**Comment by raquelalvarezbanos** [comment]
Agreed with scientists that this should be shipped with the next release, even if some of them didn't have time to test the changes.

**Comment by PranavBahuguna** [comment]
I have followed through many (though not all) steps involved in reviewing this PR. See my line notes for any issues or errors I found. So far I have:
- Checked if each new algorithm has its own unit test.
- Checked if both old and new algorithm unit tests are running the correct versions of the algorithms.
- Checked if any functionality was altered in the version 1 algorithms.
- Checked if the tests in the .rst files were running the correct versions of the algorithms.
- Checked if the system tests in `RROAutoFunctionalityTests` have been moved correctly.
- Checked if current system tests are running version 1 of the algorithms.
- Checked if the old interface is running version 1 of the algorithms.
- Checked if the new interface is running version 2 of the algorithms.

I have yet to do the following:
- Determine whether or not the algorithms in the interface reduce data correctly,
- Find how accurate the documentation for each algorithm is (and release notes).
- Code review.

File: MantidQt/CustomInterfaces/src/Reflectometry/ReflSettingsPresenter.cpp [hunk 1]
```diff
@@ -360,7 +360,8 @@ void ReflSettingsPresenter::getInstDefaults() {
 /** Generates and returns an instance of the ReflectometryReductionOne algorithm
 */
 IAlgorithm_sptr ReflSettingsPresenter::createReductionAlg() {
-  return AlgorithmManager::Instance().create("ReflectometryReductionOneAuto");
+  return AlgorithmManager::Instance().create("ReflectometryReductionOneAuto",
+                                             1);
 }
```
Comment (PranavBahuguna) [inline MantidQt/CustomInterfaces/src/Reflectometry/ReflSettingsPresenter.cpp]: Would this not run the version 1 algorithm?

**Comment by PranavBahuguna** [comment]
Other issues:
- `CreateTransmissionWorkspaceAutoTest::test_exec()` - on line 79 it uses the version 2 algorithm, rather than version 1.

File: MantidQt/CustomInterfaces/src/Reflectometry/ReflSettingsPresenter.cpp (hunk 1 above)
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/src/Reflectometry/ReflSettingsPresenter.cpp]: Yes, that should be changed.

File: docs/source/algorithms/ReflectometryReductionOne-v2.rst [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+converted to wavelength (note that :literal:`AlignBins` will be set to :literal:`True` for this in 
+:ref:`algm-ConvertUnits`) and the detectors
+of interest, specified via :literal:`ProcessingInstructions`, will be grouped together using
+:ref:`algm-GroupDetectors`. **Optionally**, the algorithm will perform direct beam
+normalization (if :literal:`RegionOfDirectBeam` is specified) dividing the detectors of
+interest by the direct beam, and monitor normalization (if :literal:`I0MonitorIndex` and
+:literal:`MonitorBackgroundWavelengthMin` and :literal:`MonitorBackgroundWavelengthMax` are all specified),
+in which case the detectors of interest will be divided by monitors. Detectors can be normalized
+by integrated monitors by setting :literal:`NormalizeByIntegratedMonitors` to true, in which case
+:literal:`MonitorIntegrationWavelengthMin` and :literal:`MonitorIntegrationWavelenthMax` will
```
Comment (PranavBahuguna) [inline docs/source/algorithms/ReflectometryReductionOne-v2.rst]: Typo - "MonitorIntegrationWavelengthMax"

File: docs/source/algorithms/ReflectometryReductionOne-v2.rst [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+workspace. If spectrum numbers do not match, the algorithm will throw and exception
+and execution of the algorithm will be stopped. This behaviour can be optionally
+switched off by setting :literal:`StrictSpectrumChecking` to false, in which case
+a warning message will be shown instead.
+
+When normalizing by transmission run, this algorithm will run
+:ref:`algm-CreateTransmissionWorkspace` as a child algorithm, with properties :literal:`WavelengthMin`,
+:literal:`WavelengthMax`, :literal:`I0MonitorIndex`, :literal:`MonitorBackgroundWavelengthMin`,
+:literal:`MonitorBackgroundWavelengthMax`, :literal:`MonitorIntegrationWavelengthMin`,
+:literal:`MonitorIntegrationWavelengthMax`, and :literal:`ProcessingCommands`.
```
Comment (PranavBahuguna) [inline docs/source/algorithms/ReflectometryReductionOne-v2.rst]: The parameter is "ProcessingInstructions", not "ProcessingCommands"

File: docs/source/algorithms/ReflectometryReductionOneAuto-v2.rst [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+Polarization Analysis Off
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If :literal:`PolarizationAnalysis = None` the reduction stops here. Note that if
+transmission runs are given in the form of a workspace group, each member in the
+transmission group will be associated to the corresponding member in the input
+workspace group, i.e., the first item in the transmission group will be used as the
+transmission run for the first workspace in the input workspace group, the second
+element in the transmission group will be used as the transmission run for the
+second workspace in the input workspace group, etc. This is also illustrated
```
Comment (PranavBahuguna) [inline docs/source/algorithms/ReflectometryReductionOneAuto-v2.rst]: I feel that it should also be mentioned that if only one item is given for 1st/2nd transmission run, that run would be used for all members of the input workspace group.

File: docs/source/algorithms/CreateTransmissionWorkspaceAuto-v2.rst [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+-----------
+
+This algorithm is a facade over :ref:`algm-CreateTransmissionWorkspace`. It pulls numeric parameters
+out of the instrument parameter file where possible. You can override any of these automatically applied
+defaults by providing your own values.
+
+The first transmission run is the only mandatory property. If :literal:`ProcessingInstructions` is not set its value is inferred from other properties:
+
+* If :literal:`AnalysisMode = PointDetectorAnalaysis` and :literal:`PointDetectorStart = PointDetectorStop` then :literal:`ProcessingInstructions` is set to :literal:`PointDetectorStart`.
+* If :literal:`AnalysisMode = PointDetectorAnalaysis` and :literal:`PointDetectorStart != PointDetectorStop` then :literal:`ProcessingInstructions` is set to :literal`PointDetectorStart:PointDetectorStop`.
```
Comment (PranavBahuguna) [inline docs/source/algorithms/CreateTransmissionWorkspaceAuto-v2.rst]: Typo - :literal:`PointDetectorStart:PointDetectorStop` - it displays incorrectly

File: docs/source/algorithms/CreateTransmissionWorkspaceAuto-v2.rst [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+* If :literal:`AnalysisMode = MultiDetectorAnalaysis` then the spectrum range from :literal:`MultiDetectorStart` to the last spectrum in the input workspace is used.
+
+When :literal:`WavelengthMin` and/or :literal:`WavelengthMax` are not provided, the algorithm will
+search for :literal:`LambdaMin` and :literal:`LambdaMax` in the parameter file. If monitor properties,
+:literal:`I0MonitorIndex`, :literal:`MonitorBackgroundWavelenghtMin`, :literal:`MonitorBackgroundWavelengthMax`,
```
Comment (PranavBahuguna) [inline docs/source/algorithms/CreateTransmissionWorkspaceAuto-v2.rst]: Typo - "MonitorBackgroundWavelengthMin"

File: docs/source/algorithms/ReflectometryReductionOneAuto-v2.rst [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+workspace group, :literal:`[1]` the second element, etc.
+
+Polarization Analysis On
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+If :literal:`PolarizationAnalysis` is set to :literal:`PA` or :literal:`PNR`
+the reduction continues and polarization corrections will be applied to
+the output workspace in wavelength. The algorithm will use the properties :literal:`PolarizationAnalysis`,
+:literal:`CPp`, :literal:`CAp`, :literal:`CRho` and :literal:`CAlpha` to run :ref:`algm-PolarizationCorrection`.
+The result will be a new workspace in wavelenght, which will override the previous one, that will
```
Comment (PranavBahuguna) [inline docs/source/algorithms/ReflectometryReductionOneAuto-v2.rst]: Typo - "wavelength"

**Comment by PranavBahuguna** [comment]
I have finished testing the documentation and interface as well as running your script. The documentation is quite accurate and describes what happens well, though I've added some line comments for mistakes and parts I thought needed more explanation on.

I've run your script and the plotted the results - they mostly match very well so there is no issues there either.

I have also tested both the new and old interfaces using `INTER_NR_test2.tbl` from the sample ISIS dataset. While the new interface seems to reduce the whole dataset just fine, the old Reflectometry interface throws an error with `CropWorkspace` - "StartSpectrum out of range!". Could you double check to see if you're having a similar problem with this?

**Comment by raquelalvarezbanos** [comment]
Thanks @PranavBahuguna, I'll add the changes you've mentioned.

**Comment by PranavBahuguna** [comment]
All looks good now :shipit:

**Comment by martyngigg** [comment]
@raquelalvarezbanos There have been some conflicts in the release notes.

**Comment by raquelalvarezbanos** [comment]
Thanks @martyngigg I'll fix them.

**Comment by martyngigg** [comment]
Test failure is unrelated to these changes.

**Pull request merged**